src/api/cpp/cvc5.cpp

```diff
@@ -3564,7 +3564,27 @@ DatatypeDecl::~DatatypeDecl()
 
 bool DatatypeDecl::isResolved() const
 {
-  return d_dtype == nullptr || d_dtype->isResolved();
+  if (d_dtype == nullptr)
+  {
+    return true;
+  }
+  // We are resolved if a constructor is resolved. Note that since
+  // internal::DType objects are copied in Solver::mkDatatypeSorts, the
+  // constructors of d_dtype are passed to NodeManager but not d_type itself.
+  // Thus, we must check whether our constructors are resolved.
+  // This is a workaround; a clearer implementation would avoid the
+  // copying of DType in Solver::mkDatatypeSorts.
```

Comment:
Do you plan to refactor this in the future? If so, maybe we should open an issue for this.

Reply:
Done